rust_src/build.rs

```diff
@@ -0,0 +1,206 @@
+                if line.starts_with("#[lisp_fn") {
+                    if let Some(begin) = line.find(C_NAME) {
+                        let input = line.to_string();
+                        let start = begin + C_NAME.len();
+                        let end = &input[start..].find('"').unwrap() + start;
+                        let name = (&line[start..end]).to_string();
+                        // Ignore macros, nothing we can do with them
+                        if !name.starts_with('$') {
+                            // even though we do not need to parse the
+                            // next line of the file this keeps all of the
+                            // lisp_fn code in one place.
+                            parse_state = ParseState::LispFn(Some(name));
+                        }
+                    } else {
+                        parse_state = ParseState::LispFn(None);
+                    }
+                    has_lisp_fns = true;
+                } else if line.starts_with("#[no_mangle]") {
+                    parse_state = ParseState::NoMangleFn;
+                } else if line.starts_with("/*") {
+                    if !line.ends_with("*/") {
+                        parse_state = ParseState::IgnoreComments;
+                    }
+                } else if line.starts_with("include!(concat!(env!(\"OUT_DIR\"),") {
+                    has_include = true;
+                }
+            }
+
+            ParseState::IgnoreComments => if line.starts_with("*/") || line.ends_with("*/") {
+                parse_state = ParseState::Looking;
+            },
+
+            ParseState::LispFn(name) => {
+                if line.starts_with("pub") || line.starts_with("fn") {
+                    if let Some(func) = name.or_else(|| get_function_name(&line)) {
+                        write!(out_file, "pub use {}::F{};\n", modname, func)?;
+                        exported.push(func);
+                    }
+                }
+
+                parse_state = ParseState::Looking;
+            }
+
+            // export public #[no_mangle] functions
+            ParseState::NoMangleFn => {
+                if line.starts_with("pub") {
+                    if let Some(func) = get_function_name(&line) {
+                        write!(out_file, "pub use {}::{};\n", modname, func)?;
+                    }
+                }
+
+                parse_state = ParseState::Looking;
+            }
+        };
+    }
+
+    if has_lisp_fns {
+        let path =
+            PathBuf::from(env::var("OUT_DIR").unwrap()).join([modname, "_exports.rs"].concat());
+        let mut exports_file = File::create(path)?;
+
+        write!(
+            exports_file,
+            "export_lisp_fns! {{ {} }}",
+            exported.join(", ")
+        )?;
+
+        if !has_include {
+            panic!(
+                [
+                    modname,
+                    ".rs is missing the required include for lisp_fn exports"
+                ].concat()
```

Comment:
seeing the concat here is kind of strange. `format!()` is more idiomatic I think.